### PR TITLE
ZK-4771: Provide a way to add theme-uri without extends any existing …

### PR DESCRIPTION
**Pull request by plijyfes**
…ThemeProvider

### CODE REVIEW

File: zk/src/org/zkoss/zk/ui/util/Configuration.java [hunk 1]
```diff
@@ -1324,6 +1325,29 @@ public void addDisabledThemeURI(String uri) {
 		return Collections.emptySet();
 	}
 
+	/**
+	 * Add a {@link ThemeURIHandler}.
+	 *
+	 * <p>Note: if {@link ThemeProvider} is specified ({@link #setThemeProvider}),
+	 * the final theme URIs generated depend on {@link ThemeProvider#getThemeURIs}.
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/util/Configuration.java:1332]: generated theme URIs

File: zk/src/org/zkoss/zk/ui/util/ThemeURIModifier.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * @author leon
+ * @since 9.6.0
+ */
+public interface ThemeURIModifier {
+	/**
+	 * Returns a read-only list of theme uri which present the current status of the CSS/WCS files list.
+	 * @return a read-only list of StyleSheet
+	 */
+	public List<StyleSheet> getReadOnlyURIs();
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/util/ThemeURIModifier.java:30]: getURIs()

File: zk/src/org/zkoss/zk/ui/util/ThemeURIHandler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+Copyright (C) 2021 Potix Corporation. All Rights Reserved.
+*/
+
+package org.zkoss.zk.ui.util;
+
+import org.zkoss.zk.ui.Execution;
+
+/**
+ * Used to add theme uri dynamically without extends any existing ThemeProvider.
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/util/ThemeURIHandler.java:18]: extending

File: zkdoc/release-note [hunk 4]
```diff
@@ -3,6 +3,7 @@ ZK 9.6.0
   ZK-4761: MVVM provide option to disable viewmodel annotation caches
   ZK-4330: provide select-all checkbox in tree component
   ZK-4524: blocking all user keyboard input during processing message showing
+  ZK-4771: Provide a way to add theme-uri without extends any existing ThemeProvider
```
Comment (scribetw) [inline zkdoc/release-note:6]: docs should be ready

File: zktest/src/archive/test2/config.properties [hunk 5]
```diff
@@ -3506,6 +3506,7 @@ F86-ZK-4235.zul=A,E,datefmt,library-property
 ##manually##F96-ZK-4761.zul=A,E,MVVM,MethodCache
 ##zats##F96-ZK-4330.zul=A,E,Tree,SelectAll
 ##zats##F96-ZK-4524.zul=A,E,processMask,showBusy,clearBusy
+##manually##F96-ZK-4771.zul=A,E,ThemeProvider,ThemeURI,ThemeURIHandler
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties:3509]: can be ZATS

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zk/src/org/zkoss/zk/ui/util/ThemeURIHandler.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * be called before {@link ThemeProvider} to allow developer to add CSS/WCS files.
+ * The earlier the ThemeURIHandler being added to Configuration is processed earlier.
+ * (first-in-first-out).
+ *
+ * Note: {@link ThemeProvider} still has the final decision.
+ *
+ * @author leon
+ * @since 9.6.0
+ */
+public interface ThemeURIHandler {
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/util/ThemeURIHandler.java:31]: add a note about it's always be executed.

**Review by scribetw — CHANGES_REQUESTED** [review]

**Review by DevChu — APPROVED** [review]

**Review by scribetw — APPROVED** [review]

**Pull request merged**